### ISSUE TITLE
Preview controllers fail with 500 if the user is not signed in

### ISSUE THREAD

**Issue by drusepth** [thredded/thredded#779]
Preface: My hunch is that this may be (and very well may very likely be) related to Heroku in some way. I thought it occurred when Heroku did a daily restart of their dynos, but I finally checked timestamps and doesn't seem to have any relation between Heroku restarts.

Issue: Once (or sometimes twice) per day, I get hundreds of the following error over a span of a few minutes. Immediately after, it seems to resolve itself. I haven't seen it personally on the front-end, but my guess is that users are probably seeing a 500 error when this happens.

![2018-12-02-121423_1562x502_scrot](https://user-images.githubusercontent.com/538235/49343298-3ac84f00-f62c-11e8-956b-7bd90e27df97.png)

I'm curious if anyone has any ideas on why this is happening, and potentially what I could do to further diagnose and/or fix the issue. 

Here's the full stack trace:

```
…rd-5.2.0/lib/active_record/associations/association.rb: 238:in `raise_on_type_mismatch!'
…b/active_record/associations/belongs_to_association.rb:  21:in `replace'
…lib/active_record/associations/singular_association.rb:  17:in `writer'
…/lib/active_record/associations/builder/association.rb: 116:in `user='
…vemodel-5.2.0/lib/active_model/attribute_assignment.rb:  51:in `public_send'
…vemodel-5.2.0/lib/active_model/attribute_assignment.rb:  51:in `_assign_attribute'
…vemodel-5.2.0/lib/active_model/attribute_assignment.rb:  44:in `block in _assign_attributes'
…vemodel-5.2.0/lib/active_model/attribute_assignment.rb:  43:in `each'
…vemodel-5.2.0/lib/active_model/attribute_assignment.rb:  43:in `_assign_attributes'
…record-5.2.0/lib/active_record/attribute_assignment.rb:  23:in `_assign_attributes'
…vemodel-5.2.0/lib/active_model/attribute_assignment.rb:  35:in `assign_attributes'
….5.0/gems/activerecord-5.2.0/lib/active_record/core.rb: 314:in `initialize'
…ms/activerecord-5.2.0/lib/active_record/inheritance.rb:  66:in `new'
…ms/activerecord-5.2.0/lib/active_record/inheritance.rb:  66:in `new'
…ntrollers/thredded/private_post_previews_controller.rb:   9:in `preview'
…0/lib/action_controller/metal/basic_implicit_render.rb:   6:in `send_action'
…/gems/actionpack-5.2.0/lib/abstract_controller/base.rb: 194:in `process_action'
…ionpack-5.2.0/lib/action_controller/metal/rendering.rb:  30:in `process_action'
…/actionpack-5.2.0/lib/abstract_controller/callbacks.rb:  42:in `block in process_action'
…ms/activesupport-5.2.0/lib/active_support/callbacks.rb: 132:in `run_callbacks'
…/actionpack-5.2.0/lib/abstract_controller/callbacks.rb:  41:in `process_action'
…actionpack-5.2.0/lib/action_controller/metal/rescue.rb:  22:in `process_action'
…k-5.2.0/lib/action_controller/metal/instrumentation.rb:  34:in `block in process_action'
…ctivesupport-5.2.0/lib/active_support/notifications.rb: 168:in `block in instrument'
…5.2.0/lib/active_support/notifications/instrumenter.rb:  23:in `instrument'
<truncated 74 additional frames>
…ruby/2.5.0/gems/rack-2.0.6/lib/rack/method_override.rb:  22:in `call'
…/bundle/ruby/2.5.0/gems/rack-2.0.6/lib/rack/runtime.rb:  22:in `call'
…2.5.0/gems/rack-timeout-0.5.1/lib/rack/timeout/core.rb: 123:in `block in call'
…rack-timeout-0.5.1/lib/rack/timeout/support/timeout.rb:  19:in `timeout'
…2.5.0/gems/rack-timeout-0.5.1/lib/rack/timeout/core.rb: 122:in `call'
…ctive_support/cache/strategy/local_cache_middleware.rb:  29:in `call'
…npack-5.2.0/lib/action_dispatch/middleware/executor.rb:  14:in `call'
…ionpack-5.2.0/lib/action_dispatch/middleware/static.rb: 127:in `call'
…bundle/ruby/2.5.0/gems/rack-2.0.6/lib/rack/sendfile.rb: 111:in `call'
…dle/ruby/2.5.0/gems/railties-5.2.0/lib/rails/engine.rb: 524:in `call'
…le/ruby/2.5.0/gems/railties-5.2.0/lib/rails/railtie.rb: 190:in `public_send'
…le/ruby/2.5.0/gems/railties-5.2.0/lib/rails/railtie.rb: 190:in `method_missing'
…/ruby/2.5.0/gems/puma-3.12.0/lib/puma/configuration.rb: 225:in `call'
…/bundle/ruby/2.5.0/gems/puma-3.12.0/lib/puma/server.rb: 658:in `handle_request'
…/bundle/ruby/2.5.0/gems/puma-3.12.0/lib/puma/server.rb: 472:in `process_client'
…/bundle/ruby/2.5.0/gems/puma-3.12.0/lib/puma/server.rb: 332:in `block in run'
…le/ruby/2.5.0/gems/puma-3.12.0/lib/puma/thread_pool.rb: 133:in `block in spawn_thread'
```

I'm using Thredded 0.15.3 (and will try with the latest 0.16.1 shortly, but didn't see anything in the changelogs that was explicitly related; I've made a few aesthetic view changes I need to verify still work as I upgrade). I'm also using Devise 4.4.3 for user authentication and Rails 5.2.0.

This error also doesn't seem to be contained to a particular page (i.e. it's happened on previewing new threads, viewing threads, editing replies, etc).

**Comment by glebm** [comment]
On our side, we are missing `before_action :thredded_require_login!` in the preview controller.

On your side, looks like you are not storing session data in a persisted storage, so whenever your server restarts all your users get logged out, sometimes mid-session (then they type something into the form, and the preview controller gets called without a signed in user).

**Comment by drusepth** [comment]
Ah, that's a good call on session storage. I'll get that fixed up; thanks.

**Comment by glebm** [comment]
@drusepth Fix released in v0.16.3!

**Comment by drusepth** [comment]
Awesome, thank you!
